### DISCUSSION TITLE
Proof of Attendance Infrastructure

### DISCUSSION THREAD

**Discussion by jonathanlei** [XRPLBounties/Proposals#19] [Open Bounties]
### Discussion thread for bounty: [Proof of Attendance Infrastructure](https://github.com/XRPLBounties/Proposals/blob/main/bounties/0019%20Proof%20of%20Attendance%20Infrastructure.md) ([PR](https://github.com/XRPLBounties/Proposals/pull/23))

Create a UI / library for event organizer to mint and distribute Proof of Attendance NFTs (with event metadata - <date, description, cover art...> embedded in memo field)  on the XRP Ledger. Also, enable them to track attendees, attendance history and distribute rewards.

This bounty is currently open. Discuss below!

**Comment by spicycheeto** [comment]
Working on this now. However, the NFT-devnet credential generator button is no longer working. Is there a problem with the server? 

Also, just to be clear, milestone #1 requires a class with the necessary functions to implement features used by the community event manager / tool?.

**Reply by connorjchen** [reply]
Hi, we believe the certs have expired for the NFT Devnet. We are working on it right now. In the meantime, feel free to use the regular Devnet (it has NFTs enabled, so it shouldn't be too different).

Update: NFT Devnet credential generation is now fixed.

**Reply by jonathanlei** [reply]
To answer your second question: A class would be a good option in terms of implementation. Please make sure there is sufficient documentation explaining each function.

**Reply by spicycheeto** [reply]
Sorry, I haven't had much time to work on this. I will have my version of the API finished on the 31st. 

Here is the current status: 

https://github.com/spicycheeto/XRPL-Proof-of-Attendance-Infrastructure/blob/main/README.md

Any input from you guys would be appreciated. Is my functionality right? what else could I add? Is this what your looking for?

**Comment by kuyawa** [comment]
Is this bounty still open? I'd like to make this proposal and see if it fits the requirements:

NFT TICKETS:

Requirements:
- Mint NFT tickets for events
- ie. Mint 10.000 nfts for a concert and sell them at diff prices
- People can buy one or many
- People can transfer them to friends/relatives/resell
- Tickets can have numbered seats
- Tickets can have price class like VIP, executive, business, coach, etc
- Tickets can be verified on entrance using qrcode, nfc, manual entry

Key features:
- NFT unique
- anti-fraud
- transferable
- handicap enabled
- mint on demand
- fees on event registration + tickets sold
- valid for X days (disney, conferences, etc)
- multiuser (buy 5 tix for a family in one wallet, five people enter at once with one validation)
- cancellation: refundable or reusable (future date)
- templates for tickets
- batch mint/transfer
- batch seating: section A seats 1..20, section B seats 21..30

V2.0
- printable for an extra fee
- discounts on qty sold
- discount coupons
- pre-sale
- artist tour, diff venues/dates
- venue map seat selection
- mobile wallet apple/android 

It can be done for the budget available as V1.0
The rest can be done in Version 2.0

ETC V1.0: 1 month
ETC V2.0: 2 months

**Reply by jonathanlei** [reply]
Hi, the scope provided here (NFT ticketing for sale) is beyond the bounty scope (Proof of attendance token for free), but NFT ticketing is a valid potential use case of XRPL and NFT's, feel free to apply to the next wave of XRPL grants [https://xrplgrants.org/](url) when the next wave opens if you have a more fleshed out idea.

**Reply by kuyawa** [reply]
Thanks, I'll stick to the basic requirements for attendance taking poap.xyz as example, definitely much easier

**Comment by spicycheeto** [comment]
Hello community, I have made a submission to milestone 1 for the proof of attendance infrastructure. Waiting on feedback before moving forward with other milestones and additional proposals.

Milestone 1 submission: https://github.com/spicycheeto/XRPL-NFT-Proof-of-Attendance-Infrastructure

Hopefully, this is a good start and meet expectations. There is still a lot more to do. I could use some help as well if anyone wants to group up in a team.

**Reply by jonathanlei** [reply]
Thanks for your submission! I will review this and give feedbacks within a week.

**Reply by tadejgolobic** [reply]
In my opinion, this is quite an issue: https://github.com/spicycheeto/XRPL-NFT-Proof-of-Attendance-Infrastructure/blob/2bec43614478d943a8d02ae16f471d41565fa566/backend/routes/nftRoutes.js#L44

Here, you expect that wallet seed / secret is passed to your backend service. In my opinion, this is a no go. I assume that nobody will want to pass wallet seed to your backend.

**Comment by spicycheeto** [comment]
Perhaps I could have an account hard coded on the backend for whomever runs
the server. This was a design decision I made to complete milestone 1.

I was thinking of when someone comes to collect the NFT. They will need to
pass some wallet information or account details. Looking at poap.xyz they
do it with a qr code for the xumm wallet. This passes the secret no?

Haven't got this far yet, but a hard coded account on the backend makes
sense to me. The application will need to have users and the users will
need to be charged separately or something like that in order to mint
because it charges a fee to the hard coded account.

**Reply by tadejgolobic** [reply]
If I understand this correctly, you do not need to sign the transaction. For example, if user has XUMM wallet, you can pass this sign request to XUMM, and user can then sign transaction in XUMM and give you back the payload?

@jonathanlei what do you think? Is this the correct approach?

**Reply by jonathanlei** [reply]
I agree with @tadejgolobic. XUMM is a better way to handle transaction signing without passing along secrets. You can create a payload and XUMM will return a link/QR code that you can pass back to the user to sign on their XUMM app, no secrets needed. Docs here: https://xumm.readme.io/

**Reply by jonathanlei** [reply]
@spicycheeto other than the wallet signing issue everything looks good so far, love the detailed documentations. 
Frontend is starting to look good too.

**Reply by jonathanlei** [reply]
@spicycheeto one more thing: for minting multiple [NFT's,](https://github.com/spicycheeto/XRPL-NFT-Proof-of-Attendance-Infrastructure/blob/7355c9ee218f0bd8f7c1d82819f79b1fd0d005bf/backend/routes/XrplNFTHelper.js#L107) it is best to follow the batch minting guide listed here:  https://xrpl.org/batch-minting.html

**Reply by spicycheeto** [reply]
> @spicycheeto one more thing: for minting multiple [NFT's,](https://github.com/spicycheeto/XRPL-NFT-Proof-of-Attendance-Infrastructure/blob/7355c9ee218f0bd8f7c1d82819f79b1fd0d005bf/backend/routes/XrplNFTHelper.js#L107) it is best to follow the batch minting guide listed here: https://xrpl.org/batch-minting.html

Fixed this. The rest is coming this weekend with final documentation and uploads. WAAY faster to do it this way. Great suggestion. Thank you.

**Reply by jonathanlei** [reply]
Please also add tests, as it's required in milestone 1, thanks!

**Comment by kuyawa** [comment]
I'd like to be considered for the bounty, here is my project https://kollek.me
What are the steps to participate?

**Reply by jonathanlei** [reply]
hey @kuyawa, thanks for the submission
1. to pick up/ claim a bounty - please follow instructions listed in https://github.com/XRPLBounties/Proposals/blob/main/README.md
2. just checked out your repo: frontend looks good so far, and backend needs more documentation and tests - since the milestone 1 is an API (https://github.com/XRPLBounties/Proposals/blob/main/bounties/0019%20Proof%20of%20Attendance%20Infrastructure.md)

**Reply by kuyawa** [reply]
Ok, will work on that

**Reply by tadejgolobic** [reply]
Actually, I am a little bit puzzled by this one:
https://github.com/kuyawa/kollek/blob/88243df4ccc3ab126ea201ff93a7a5cf39844788/source/api.js#L108

I think that the whole purpose of claiming NFT is not to Mint another NFT, but to send NFT of an event, to the recipient.
For instance, @jonathanlei will create event through your platform and he wants to be able to distribute 10 NFTs. Me, as a recipient, will get (I will claim) one of this pre-minted NFTs. Is this the correct flow @jonathanlei ?

**Reply by jonathanlei** [reply]
that is correct [tadejgolobic](https://github.com/tadejgolobic), good catch!  - the event organizers should cover the tx fee for minting and transferring. Also, please add a route for uploading images to IPFS or similar decentralized file storage system.

**Reply by kuyawa** [reply]
There are two ways of minting, in bulk (pre-mint) or on demand. Minting on demand is easier, optimal and resource friendly, interested parties mint ther nfts when needed. Bulking is expensive, suboptimal, resource intensive and produces excessive waste in the form of unclaimed nfts, also they include an expensive delivery process, mailing lists, etc. Imagine pre-minting 1000 nfts and only 300 are claimed. The process of pre-minting, claiming and transfering is more complex than just minting on demand

Both options have their use, I decided to offer both in my proposal

**Reply by jonathanlei** [reply]
After some research, we agree that minting on-demand is a better approach considering bulk pre-minting with no-code would be impractical for event organizers (signing hundreds of requests on their XUMM wallet). However, the mint function should including event organizer account as the minter and the event attendee as the issuer - meaning that the organizer is minting on the attendee's behalf and paying for the tx fee. For more details, check out the docs - https://xrpl.org/nftokenmint.html#issuing-on-behalf-of-another-account

**Reply by kuyawa** [reply]
That's exactly what I do in my proposal, take a look https://github.com/kuyawa/kollek-api/blob/main/kollek-api.js#L18

**Reply by tadejgolobic** [reply]
I am also working on this bounty, and I did a little bit different than you @kuyawa .
In your API for example, you can get account_info of this issuer address and check if your account is set as NFTokenMinter. If it is not, you can create new XUMM payload, so that your account is added as NFTokenMinter for provided issuer address and user will approve/deny this request on XUMM.

If you are already NFTokenMinter for the issuer, then you can just mint NFTs, with provided issuer address as issuer, and then issuer does not need to sign anything.

I personally believe that you should also include this functionality to set your account as NFTokenMinter for the provided issuer address. But that is not up to me but up to @jonathanlei :)

**Reply by kuyawa** [reply]
@tadejgolobic that's exactly how it should be done. I didn't include that part as the API should be wallet agnostic. The app should do all that in the UI, check user info, tell user if they should add the minter (a little weird since then they would be restricted to mint in other platforms) and provide a payload to add the minter. Then create a token with issuer account.

All in all, I prefer minting on demand. Instead of the rain dance, the user just mints their own NFT without any further steps. Simple always wins in my book

**Reply by jonathanlei** [reply]
@kuyawa okay I didn't see that line where issuer is provided - I think there will be some confusion for people using the API for the distinction between minter and issuer: for minting on demand, documentation wise we should clarify that minter is the event organizer who pays for the minting and tx fee and the issuer (optional) is the recipient of the POAP (event attendee). 
@tadejgolobic hopefully that provides some clarifications to your question. Also, if you have your WIP repo, you can post it here and I can take a look at your implementation. 
We do accept multiple submissions for this bounty, given they have correct implementations, documentations and tests.

**Reply by golobitch** [reply]
@jonathanlei ok, I get your point, but what about if the issuer is not provided through an API call? What about then? Should we just mint the token? We can, that is not a problem, but how would then claim work?

**Reply by jonathanlei** [reply]
Yeah if the issuer is not provided, then the organizer is minting and owning the POAP. 
For claiming, the claimer can create a buy offer with the NFT tokenID using `NFTokenCreateOffer` to the owner for the price of 0, and the owner accepts. This would be best to do through XUMM payload for the claimer.

**Reply by jonathanlei** [reply]
However, the claiming method of minting on behalf of claimer is highly encouraged, since the process is easier and no offers need to be created.

**Reply by tadejgolobic** [reply]
That is actually not possible right? Currently it is not possible to create buy offer using `NFTokenCreateOffer` and amount 0. XRPL would throw an error. Amount must be greater than 0.

> The amount must be non-zero, except where this is an offer to sell and the asset is XRP; then, it is legal to specify an amount of zero

In case you do sign and submit such transaction, you will get `temBAD_AMOUNT`.

> The Amount field is not valid. For example, the amount was zero for a buy offer, 

Source: https://xrpl.org/nftokencreateoffer.html

**Reply by jonathanlei** [reply]
You can specify the currency to XRP and set the price to 0 -

> 
Amount	[Currency Amount](https://xrpl.org/basic-data-types.html#specifying-currency-amounts)	Amount	Indicates the amount expected or offered for the corresponding NFToken. The amount must be non-zero, except where this is an offer to sell and the asset is XRP; then, it is legal to specify an amount of zero, which means that the current owner of the token is giving it away, gratis, either to anyone at all, or to the account identified by the Destination field.

Source: https://xrpl.org/nftokencreateoffer.html#nftokencreateoffer-fields

**Reply by golobitch** [reply]
Yes, but this is for sell offer. Above, you were talking about buy offer :)

**Reply by jonathanlei** [reply]
@golobitch oops that is right my bad on misreading it :) 
Then creating a sell offer with a price of 0 to the claimer and having the claimer accept would be the way to go. Unless you have other suggestions.

**Comment by kuyawa** [comment]
Here is the source https://github.com/kuyawa/kollek
Will keep working on the details

**Comment by spicycheeto** [comment]
Hey team, 

I've implemented the batch NFT minting on a node.js backend per your requests. I will have the XUMM wallet method of collecting by Sunday. Is there anything else I need to do to complete milestone 1? what are the next steps? I have additional proposals i'd like to make but want to get paid for milestone 1 first.

**Reply by spicycheeto** [reply]
Please, my final updates will all be done this weekend. I have a lot of code clean up and documentation updates to make but the batchx method (function) should be working. Also, please note, this is just the backend. I'm no longer keeping an updated version of my frontend. This will be part of a new proposal.

**Reply by jonathanlei** [reply]
Thank you for implementing the changes, according to milestone 1, tests are needed to ensure the API is functional. Also, please clean up some of the comments (todo's) and add complete documentation (params to routes and functions) to make sure people using it can have clear instructions.

**Comment by kuyawa** [comment]
Here is my proposal for the API https://github.com/kuyawa/kollek-api
Here is my proposal for the UI https://github.com/kuyawa/kollek
Here is the site running in nft-devnet https://kollek.me

**Comment by tadejgolobic** [comment]
@jonathanlei how did you imagine that NFT claim will work?
For example, I have created an API, where the person who wants to claim the NFT need to provide his account and NFTokenId. After that, I create NFTokenOfferCreate, with that provided account as destination. Is this what you guys had in mind?

Because this means that basically anyone can claim NFT if it knows the NFTokenId :)

Or should event organizer somehow import wallet addresses into the API, and we create N requests with destination as this addresses?

**Comment by jonathanlei** [comment]
Additional milestone 4 ($1500) has been approved and added: Deploying a no-code tool for non-technical event organizers. 
Check out the details here: https://github.com/XRPLBounties/Proposals/blob/main/bounties/0019%20Proof%20of%20Attendance%20Infrastructure.md

Milestones 1-3 must be completed and approved before claiming milestone 4

**Reply by spicycheeto** [reply]
I'm re-writing my API to work with Xumm. Any idea on how to do a bulk mint request to Xumm api?

edit: The only way I see to implement this is to have a master minter account owned by the site which mints on behalf of the user and issues to their account...with bulk requests.

**Reply by jonathanlei** [reply]
Yes having a central minter account would be fine!

**Comment by JustAnotherDevv** [comment]
I have created a basis for the proof of attendance system during recent hackathon https://xrplnft.devpost.com
Here are repos for early version of the app:

- Devpost page with more information https://devpost.com/software/xrp-nft-attendence
- https://github.com/JustAnotherDevv/XRPledgerAttendance
- https://github.com/JustAnotherDevv/xrpl_poa_nft_hackathon_server

Currently I am wondering if this project could be accepted for the bounty once I spend more time on fixing few issues:

- There was a problem with using XRPL library on the frontend with Vite so currently the account's seed phrase is passed to backend to transfer NFT.
- Some parts of the code needs rewriting, wasn't focusing on that during just few days of hackathon development
- Comments providing clear description of the functions should be added.
- There was some problem with passing images to the backend  that are supposed to be uploaded to IPFS directly from UI. That could be fixed very easily.
- Landing page should be added.
- More features such as burning the NFTs that weren't redeemed after event ends, querying for the token holders, generating QR codes, integration with existing XRPL wallets etc.

**Reply by jonathanlei** [reply]
Congrats on finishing the Hackathon! This is a very interesting Proof of Concept with lots of promising starts.There are few things I've noticed that need to be changed for this to qualify for the Bounty: 
1. The backend code looks good for batch minting and distribution, however, there needs to be more documentation explaining each endpoint's utility, params needed and tests need to be added. 
2. The frontend ideally would connect to XUMM wallet for non-technical users, and the backend would produce payloads for the FE users to sign. 
3. The frontend would need better design since this will be a public facing tool for event organizers. 
4. There are two milestones (Event attendees lookup and Event NFT ownership verification) that are missing from your backend, please reference [the bounty document](https://github.com/XRPLBounties/Proposals/blob/main/open_bounties/0019%20Proof%20of%20Attendance%20Infrastructure.md)
That being said, once you finish documenting and testing the backend API, please let me know and we are happy to review it, and if it passed the quality check, we are happy to pay out the first milestone, thanks!

**Reply by JustAnotherDevv** [reply]
Hey @jonathanlei I have updated my project after hackathon. 
Now it includes features for first 3 milestones, the project is separated into class and example expressJS API.
There are proper comments, tests in Mocha and documentation. You can find the repo here - https://github.com/JustAnotherDevv/Attendify_XRPL.
It would be great to get your feedback if this submission meets all criteria. 
I am also wondering if it make sense to add features related to more sophisticated configuration like whitelists, deadlines after which it's not possible to claim NFT etc or should those features be added as another milestone.

**Reply by jonathanlei** [reply]
Having reviewed your bounty with @JST5000  and @ckniffen , you’ve made a lot of great progress!

We had a look and here’s what you’ll need to do for each of the milestone components:

**For milestone 1:**
- Currently it seems that this will only work with testnet, but the api needs to work with mainnet addresses as well. Can you please add a way for the minter to set their issuing account? (Maybe by pulling a secret key from an environment variable?)
    - Currently the `eventId` is the minter address which won’t work if a single minter has multiple events. You should generate a separate code per batch mint operation to allow one minter to have multiple events. 
- `getBatchNFTokens` currently could get the wrong result because it doesn’t paginate through responses from the server. Servers only send the first couple results, then if you want more you have to keep asking. More details here: https://xrpl.org/markers-and-pagination.html
- You can only have [250 active tickets](https://xrpl.org/tickets.html#limitations). The code does not take into account batch minting over that amount or if the account were to have active tickets already as could happen with concurrent requests.
    - One other thing to note is that 250 active tickets can cause a large reserve requirement (Currently [2 XRP per object](https://xrpl.org/reserves.html#:~:text=On%20Mainnet%2C%20the%20current%20minimum,set%20aside%20this%20much%20XRP.), so 500 XRP for just tickets) 
    - To deal with this, you may want to check account balances as well to make sure you don’t accidentally try to make too many objects - although that may be something you correct with documentation in the README “You must have at least X XRP in order to use this to handle the reserve requirements for the NFTs and the minting process which uses [Tickets](https://xrpl.org/tickets.html).”)
- The REST methods should be documented (preferably in the README.md) so that people who are using the repo can tell what the interface is for the endpoints. 

_Questions:_
- Why is the `minter` added to the `claimableAddresses` array? (It seems weird that the minter would be able to claim their own NFT - is there something else it’s doing?)


**For milestone 2:**
- Currently you’re using in-memory storage to keep track of who has already claimed an NFT, but if the server goes down you end up losing all of your data. So, you should use a more permanent state to check if they are able to claim the NFT to avoid duplicate claims.
    - One way you can solve this is by using the eventId as a token taxon when minting the NFT, then looking up on the ledger whether the receiving address has ever claimed / minted an NFT with that event id from your issuing account. (Basically using the ledger as your permanent state) 


**For milestone 3:**
- Currently the function `verifyOwnership` checks if someone owns a specific NFT, but for verifying if someone has any token from the event, it’s unclear how someone would get the NFTokenID to pass into this function. You probably need something which looks up the details of the NFT (ex. Location, event code, etc.) and ensures that it was issued by the correct account.
- You can use the `parseNFTokenID` from xrpl.js to get a lot of that information from the NFT itself without a network call.
Alternatively, if you can help us understand the user story for how someone who attended an event could look up the information they need to verify to someone else using this function that they actually own an NFT from an event. (We’re just a bit confused on the flow)


**Additional nits / optional suggestions:**
- Thanks for generating the docs, but they’re hard to read right now since you have to manually go through the html to read the text. It’d be very helpful to have that hosted somewhere. (This is a quick guide on how to set that up which might be helpful! https://codeburst.io/generate-docs-and-host-it-with-jsdoc-and-github-pages-113b9dae9acb)
- If you’re comfortable with it (not required) could you include the demo video showing how the tool works in the README? (We think it’d be helpful to show people the feature rather than just listing them - especially if there are barely-technical users)
- Nit: We’d recommend moving the constants in `attendify.js` into a file called `constants.js` or `utils.js` to make them easier to discover.
- Nit: In `attendify.js` (and maybe other files) you spell `signature` as `signture` which is a slight spelling mistake.
- Nit: For Flags use human readable versions from xrpl.js

We know this is a lot of feedback, so if you have any questions please let us know and we are happy to clarify / discuss what the best way to go about it could be!

**Reply by JustAnotherDevv** [reply]
**Milestone 1**
In my original implementation I wanted to generate new temporary wallet that would be used for minting and distribution of NFTs. My idea was that would make it harder to guess claim id for any particular event and it would give more agency to people creating those claim events because later on seed phrase to this new wallet could be given to them. 
Looking at your feedback now I see that it was probably overengineered and I am already modifying my submission to include single account for minting and distribution of NFTs. `claimableAddresses ` was the array storing these generated accounts that's why each `minter` was added there.

**Milestone 2**
After looking at documentation I don't see a way to query for getting just NFTs that were issued by specific account. Something similar to method `account_nfts` like 
`await client.request({
        method: "nfts_by_issuer",
        account: address,
      })`
Is it acceptable for the backend server to just store the list of participants adressess in a file/database?

**Milestone 3**
Is it acceptable that `verifyOwnership ` function would take 2 additional args `issuer` and `eventID` then it would check whether or not user requesting verification has any NFT that was issued by `issuer`  address? In this case the ID for specific NFT is no longer needed. I think that in the end the user could request verification for event ID that's stored in NFT Taxon.

Once all requested changes are completed and documented I will host the docs and demo video showcasing all of the features.

**Reply by JST5000** [reply]
Re: `Milestone 2`
1. There's this handy api you can use to get all NFTs with a given issuer :) - https://api.xrpldata.com/docs/static/index.html (Second command in the list). 
2. Storing in a file/database seems fine to me :)

Re: `Milestone 3`
1. I think that makes sense. So if a user has an NFT, they have to look up the `Issuer` and `NFTokenTaxon` then they can provide those to the API in some way along with a signature, and it'll show they are in fact the owner of that NFT (is that right?)

**Comment by jonathanlei** [comment]
Announcement: We have a new addition to the process that will make it easier for reviewers and submitters to collaborate. Going forward, you will share your submission by opening a PR to new open source repo(s) under the [XRPL Bounties github organization](https://github.com/XRPLBounties).

This will make it easier for reviewers and submitters to collaborate (comment on specific line of codes and request changes). It will also increase the publicity of your work, as the community will be able to find everything under XRPL Bounties.

What to do: Reply here with a link to your work-in-progress repo. If you've made enough progress for us to start requesting revisions, we will create an open source repo (under the XRPLBounties github org) for you to open PR(s) against.
The repo's default branch will start off empty, so that your PR will include your entire submission. This allows reviewers to easily comment directly on specific lines of code and collaborate with you more easily with PR reviews.

From now on, this process will be implemented for all bounties (except ones that are submitted to repos within [XRPLF](https://github.com/XRPLF)).
Thanks!

**Comment by jonathanlei** [comment]
The repository for this bounty has been created at: https://github.com/XRPLBounties/Proof-of-Attendance,
To be considered for the bounty, please create a PR against the main branch, thanks!

**Reply by JST5000** [reply]
(For context - the reason we're asking for that is so we can provide in-line comments on code & you can have more transparency into our approval process)

**Reply by JustAnotherDevv** [reply]
I have created PR to the repo with updated codebase. Will continue posting updates in there just like you requested.

**Comment by spicycheeto** [comment]
Resolved and did updates for all comments on Proof of Attendance submission #2. Awaiting further review and any additional comments / requirements.

**Comment by jonathanlei** [comment]
Milestone 1-3 has been claimed by @JustAnotherDevv, link to the API is here: https://github.com/XRPLBounties/Proof-of-Attendance.  thanks for everyone's hard work, milestone 4 is still open.

**Comment by JustAnotherDevv** [comment]
7034

**Reply by JST5000** [reply]
Thanks for confirming your GitHub :)

**Comment by jiveyTO** [comment]
Minted 10 NFTs for attendees of our Toronto meetup in Feb

<img width="1365" alt="CleanShot 2023-03-05 at 23 14 39@2x" src="https://user-images.githubusercontent.com/20135683/223018536-ec310e91-425b-4f23-ba14-b89268e91bdc.png">

**Reply by JustAnotherDevv** [reply]
Awesome, happy to see first use case of my project. Will try to work on the official UI integration soon.
